lib/switchtec.c

```diff
@@ -652,33 +674,223 @@ int switchtec_hard_reset(struct switchtec_dev *dev)
 			     NULL, 0);
 }
 
-static int log_a_to_file(struct switchtec_dev *dev, int sub_cmd_id, int fd)
+/**
+ * @brief Reads a log definition file and stores the definitions
+ * @param[in] log_def_file - log definition file
+ * @param[out] log_defs    - log definitions
+ * @return 0 on success, negative value on failure
+ */
+static int read_log_defs(FILE *log_def_file, struct module_log_defs *log_defs)
 {
-	int ret;
+	char line[LOG_ENTRY_LEN_MAX];
+	char mod_name[FW_MOD_NAME_LEN_MAX];
+	int mod_id;
+	int num_entries;
+	int entry_idx;
+
+	while (fgets(line, sizeof(line), log_def_file)) {
+		/*
+		 * Parse the module name, module ID, and number of module log
+		 * entries
+		 */
+		if (sscanf(line, "%31s %i %i",
+			   mod_name, &mod_id, &num_entries) != 3)
+			continue;
+
+		/* validate the module ID */
+		if (mod_id > FW_MOD_ID_MAX) {
```

Comment:
I'd probably re-code this to allocate the output dynamically instead of hard coding random limits which may cause issues down the road and wastes a lot of memory. There are a number of approaches to this, depending on what the actual data looks like but I'm thinking something along the lines of using realloc... roughly (note I didn't even compile test this):

```
struct module_log_defs {
	char *mod_name;		 
	char **entries; 
	int num_entries;			
};

static void free_log_defs(struct module_log_defs *defs, int count)
{
	int i, j; 

	if (!defs)
		return;

	for (i = 0; i < count; i++) {
		free(defs[i].mod_name);
		if (!defs[i].entries)
			continue;
		for (j = 0; j < defs[i].num_entries; j++) {
			free(defs[i].entries[j]);
		}
		free(defs[i].entries);
	}
	free(defs);
}

static struct module_log_defs *realloc_log_defs(struct module_log_defs *old, 
						int num_entries, int *allocated)
{
	struct module_log_defs *new;
	int i;
        
	new = reallocarray(old, num_entries, sizeof(*new));
	if (!new) {
		free_log_defs(old, *allocated);
		return NULL;
	}
        
	for (i = *allocated; i < num_entries; i++) {
		new[i].mod_name = NULL;
		new[i].entries = NULL;
	}
           
	*allocated = num_entries;

	return new;
}

static struct module_log_defs *read_log_defs(FILE *log_def_file, log_defs)
{ 
	struct module_log_defs *defs = NULL;
	int mod_id, num_entries;  	
	int mods_allocated = 0;
	char *line, *mod_name;
	int ret, i;

	while (fscanf(log_def_file, "%ms", &line)) {
		ret = sscanf(line, "%ms %i %i", &mod_name, &mod_id, 
			     &num_entries);
		free(line);
		if (ret != 3) 
			continue;

		if (mod_id >= mods_allocated)
			defs = realloc_log_defs(defs, mod_id * 2, 
						&mods_allocated);
		if (!defs) {
			free(mod_name);
			return NULL;
		}

		defs[mod_id].mod_name = mod_name;
		defs[mod_id].num_entries = num_entries;
		defs[mod_id].entries = calloc(num_entries, 
					      sizeof(*defs->entries));
		if (!defs[mod_id].entries) {
			free_log_defs(defs, mods_allocated);
			return NULL;
		}

		for (i = 0; i < num_entrise; i++) {
			ret = fscanf(log_def_file, "%ms", 
				     &defs[mod_id].entries[i]);
			if (!ret) {
				errno = SWITCHTEC_ERR_LOGDEFS_PARSE_ERR;
				free_log_defs(defs, mods_allocated);
				return NULL;                                
			}
		}
			
	}

	return defs;
}
```

Reply:
I agree that allocating things dynamically would be nice, particularly for memory wastage. One thing I don't think we can do though is use the 'm' scanf modifier to allocate strings, as it's not portable. It doesn't even work in MSYS2.

What we could do instead is to read each line into a large fixed-size buffer. Lines would be read one at a time so we'd only need one such buffer. Then when storing the log text we can calloc each entry based on the string size.

For the module name, these are usually around 3-4 characters and typically no more than 10. A fixed 16-byte buffer should be fine, and worst case if there happened to be a huge module name we can just truncate it. Keeping it at a known fixed size also simplifies printing the heading.

I do think the number of modules and the number of entries per module should be dynamically allocated, as you're suggesting, as this could save a lot of memory. A minor thing I noticed is that your sample code is allocating room for (mod_id * 2) modules when it runs out of space. This might be excessive as the module IDs get into the hundreds. How about something like (mod_id + 20), so we're guaranteed not to over-allocate by 20 modules?

Reply:
Oh, yes, if msys2 doesn't support it then you can't use it. You'll have to have a fixed line buffer and then strdup(). Just be careful with handling a too long module-name...  This change makes the code a bit more complicated but it shouldn't be terrible.

Using *2 for reallocated arrays is pretty common. It's better to have a logarithmic scale than a linear one for reallocation. The size of the struct is also now pretty small so it's not like we are wasting much memory by having it be double the size used. The point is to not reallocate too many times because it requires a lot of memory copying. It may make sense to tweak this to a typical log file. So, if it's typically in the 100s of elements maybe have the initial allocation start at ~200ish. Test with smaller values first though.

Reply:
Sounds good -- I'll do the fixed-length buffer and strdup() approach.

I'll use *2 for the reallocation then. We currently have about 135 modules and that probably won't grow very much, so I'll start it at 200 (but test using small values).